src/Akka.Hosting/ActorRegistry.cs

```diff
@@ -84,6 +134,72 @@ public bool TryGet(Type key, out IActorRef actor)
             return false;
         }
 
+        public async Task<IActorRef> GetAsync<TKey>(CancellationToken ct)
+        {
+            return await GetAsync(typeof(TKey), ct);
+        }
+
+        public async Task<IActorRef> GetAsync(Type key, CancellationToken ct)
+        {
+            // try to get the populated actor first, if available
+            if (TryGet(key, out var storedActor))
+            {
+                return storedActor;
+            }
+
+            var tcs = new TaskCompletionSource<IActorRef>(TaskCreationOptions.RunContinuationsAsynchronously);
+            var waitingRegistration = new WaitForActorRegistration(key, tcs);
+            var registration = ct.Register(CancelWaiter(key, ct, waitingRegistration), _actorWaiters);
+            waitingRegistration.CancellationRegistration = registration;
+
+            _actorWaiters.AddOrUpdate(key,
```

Comment:
ConcurrentDictionary add semantics make this clean on additions and on removes... I think I'll clean this up for cancellations too.